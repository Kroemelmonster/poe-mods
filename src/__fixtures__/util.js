export const findBy = key => (haystack, needle) =>
  haystack.find(({ [key]: value }) => value === needle);

export const findByPrimary = findBy('primary');
