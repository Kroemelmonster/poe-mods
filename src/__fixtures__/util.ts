import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import MasterBenchOption from '../generators/MasterBenchOption';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T>(
  file: keyof Tables,
  constructor: Buildable<P, T>,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(path.join(__dirname, `./${file}.json`));

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption
  >;
  items: PropsTable<BaseItemTypeProps, Item>;
  mods: PropsTable<ModProps, Mod>;
}

export const createTables = () => {
  return {
    atlas: createTable('atlas', AtlasNode),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
    ),
    items: createTable('items', Item),
    mods: createTable('mods', Mod),
  } as Tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};
