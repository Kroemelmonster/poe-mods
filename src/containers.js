// @flow
export * from './containers/index';
