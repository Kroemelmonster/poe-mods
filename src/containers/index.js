// @flow
// api surface
export { default as AtlasNode } from './AtlasNode';
export type { Container } from './Container';
export { default as ImmutableContainer } from './ImmutableContainer';
export { default as Item } from './Item';
export { default as Implicits } from './Implicits';
export type { Rarity } from './Item';
