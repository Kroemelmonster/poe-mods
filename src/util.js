// @flow
export * from './util/index';
