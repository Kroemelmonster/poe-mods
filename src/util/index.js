// @flow
import { type Flags } from './Flags';
import { type ValueRange } from './ValueRange';

export type { Flags };
export type { ValueRange };
export { anySet } from './Flags';
export { default as Atlas } from './Atlas';
export { default as MasterBench } from './MasterBench';
export { default as Stat } from './Stat';
