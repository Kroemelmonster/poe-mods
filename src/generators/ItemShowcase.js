// @flow
import type { Item } from '../containers/';
import type { CraftingBenchOptionsProps, ModProps } from '../schema';

import { Mod } from '../mods/';
import { MasterBench } from '../util';
import Generator, { type GeneratorDetails } from './Generator';
import { Talisman, Transmute, Vaal } from './item_orbs/';

/**
 * Masterbench/Currency hybrid
 */
export default class ItemShowcase extends Generator<Mod, Item> {
  master: MasterBench;
  talisman: Talisman;
  transmute: Transmute;
  vaal: Vaal;

  constructor(props: ModProps[], options: CraftingBenchOptionsProps[]) {
    const master = MasterBench.build(options);
    const talisman = Talisman.build(props);
    const transmute = Transmute.build(props);
    const vaal = Vaal.build(props);

    const mods = [
      ...talisman.mods,
      ...transmute.mods,
      ...vaal.mods,
      ...master.getAvailableMods(),
    ];

    super(mods);

    this.master = master;
    this.talisman = talisman;
    this.transmute = transmute;
    this.vaal = vaal;
  }

  /**
   * only abstract showcase, not for actual usage
   */
  applyTo(item: Item) {
    return item;
  }

  /**
   * greps mod::applicableTo and (if implemented) mod::spawnableOn 
   * if we have all the space for mods we need
   */
  modsFor(item: Item, whitelist: string[] = []) {
    const details = [
      ...this.master.modsFor(item, whitelist),
      ...this.talisman.modsFor(item, whitelist),
      ...this.transmute.modsFor(item, whitelist),
      ...this.vaal.modsFor(item, whitelist),
    ];

    // flow cant merge object types
    // { mod: OneMod, prop: number } | { mod: AnotherMod, prop: number }
    // will not become { mod: OneMod | AnotherMod, prop: number }
    // and for some reason ['a', 'b'] cant be cast to Array<number | string>
    return ((details: any): GeneratorDetails<Mod>[]);
  }
}
