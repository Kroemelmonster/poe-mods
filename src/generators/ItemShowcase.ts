import Item from '../containers/item/Item';
import MasterBench from '../helpers/MasterBench';
import Mod from '../mods/Mod';
import { CraftingBenchOptionsProps, ModProps } from '../schema';

import Generator, { GeneratorDetails } from './Generator';
import Alchemy from './item_orbs/Alchemy';
import EnchantmentBench from './item_orbs/EnchantmentBench';
import Vaal from './item_orbs/Vaal';
import { Flags } from '../util/Flags';

/**
 * Masterbench/Currency hybrid
 */
export default class ItemShowcase extends Generator<Mod, Item> {
  public enchantment: EnchantmentBench;
  public master: MasterBench;
  public explicits: Alchemy;
  public vaal: Vaal;

  constructor(props: ModProps[], options: CraftingBenchOptionsProps[]) {
    const enchantment = EnchantmentBench.build(props);
    const master = MasterBench.build(options);
    const explicits = Alchemy.build(props);
    const vaal = Vaal.build(props);

    const mods = [
      ...enchantment.mods,
      ...explicits.mods,
      ...vaal.mods,
      ...master.getAvailableMods(),
    ];

    super(mods);

    this.enchantment = enchantment;
    this.master = master;
    this.explicits = explicits;
    this.vaal = vaal;
  }

  /**
   * only abstract showcase, not for actual usage
   */
  public applyTo(item: Item) {
    return item;
  }

  /**
   * not applicable to anything
   * @param item 
   */
  public applicableTo(item: Item): Flags {
    return {
      applicable: false,
    } as Flags;
  }

  /**
   * greps mod::applicableTo and (if implemented) mod::spawnableOn 
   * if we have all the space for mods we need
   */
  public modsFor(
    item: Item,
    whitelist: string[] = [],
  ): Array<GeneratorDetails<Mod>> {
    const details = [
      ...this.master.modsFor(item, whitelist),
      ...this.enchantment.modsFor(item, whitelist),
      ...this.explicits.modsFor(item, whitelist),
      ...this.vaal.modsFor(item, whitelist),
    ];

    // flow cant merge object types
    // { mod: OneMod, prop: number } | { mod: AnotherMod, prop: number }
    // will not become { mod: OneMod | AnotherMod, prop: number }
    // and for some reason ['a', 'b'] cant be cast to Array<number | string>
    return details;
  }
}
