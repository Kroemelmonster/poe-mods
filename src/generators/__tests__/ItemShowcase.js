// @flow
import ItemShowcase from '../ItemShowcase';
import { Item } from '../../containers/';
import { findByPrimary } from '../../__fixtures__/util';

const baseitemtypes = require('../../__fixtures__/baseitemtypes.json');
const craftingbenchoptions = require('../../__fixtures__/craftingbenchoptions.json');
const mods = require('../../__fixtures__/mods.json');

it('should build', () => {
  const showcase = new ItemShowcase(mods, craftingbenchoptions);

  expect(showcase).toBeInstanceOf(ItemShowcase);
  expect(showcase.mods.length).toBeGreaterThan(0);
});

it('should not do anything, just showcase', () => {
  const showcase = new ItemShowcase(mods, craftingbenchoptions);
  const greaves = Item.build(findByPrimary(baseitemtypes, 1650));

  expect(showcase.applyTo(greaves)).toBe(greaves);
  expect(showcase.modsFor(greaves).length).toBeGreaterThan(0);
});
