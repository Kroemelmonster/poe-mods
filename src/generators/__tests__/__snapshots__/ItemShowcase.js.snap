// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should use special rarity 1`] = `
Array [
  1033,
  1034,
  1035,
  1036,
  1037,
  1038,
  1039,
  1223,
  1224,
  1225,
  1226,
  1227,
  1228,
  1448,
  1449,
  1450,
  1491,
  1492,
  1503,
  1504,
  1505,
  1506,
  1507,
  1809,
  1810,
]
`;
